### PR TITLE
Trapping rain water

### PR DESCRIPTION
**Pull request by Mushtary**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_trapping_rain_water.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+def trap(height):
```
Comment (Ismatova-Rumiya) [inline solutions/tests/test_trapping_rain_water.py:1]: Function docstring needs to be added. Please, add detailed explanation of input, output, and logic used in the function. For example: 
"""
 Calculates the amount of rainwater that can be trapped given an elevation map.
    Input:
    height (list[int]): A list of non-negative integers representing the elevation map.”
But, you have to include for logic as well

File: solutions/trapping_rain_water.py
Comment (Ismatova-Rumiya) [inline solutions/trapping_rain_water.py:1]: As per my understanding, according to the rules provided by Evan, we are suppose to write test cases for our code and then provide docstring for each test case. Please, review this again and edit accordingly.

**Review by Ismatova-Rumiya — COMMENTED** [review]
Generally, you have written a clear code. My biggest suggestion would be paying attention to the fact that we have to write test cases for our code and then provide comment/docstring for each test case.

File: solutions/trapping_rain_water.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+def trap(height):
+    if not height or len(height) < 3:
```
[the remark this answers is not in the document]
Comment (salihakalender) [reply solutions/trapping_rain_water.py:2]: It would be beneficial to add a detailed explanation of the argument and return value in the docstring, as mentioned in the checklist. Providing this information helps future readers and collaborators understand the function’s input and output requirements more effectively.

[the remark this answers is not in the document]
Comment (salihakalender) [reply solutions/trapping_rain_water.py:2]: Please add at least 3 examples in the docstring for testing.

File: solutions/tests/test_trapping_rain_water.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    if not height or len(height) < 3:
```
[the remark this answers is not in the document]
Comment (salihakalender) [reply solutions/tests/test_trapping_rain_water.py:2]: Please add tests for defensive assertions and tests for many boundary cases, as mentioned in the checklist.

[the remark this answers is not in the document]
Comment (salihakalender) [reply solutions/tests/test_trapping_rain_water.py:2]: Boundary case handling is present in the current code, but it is not clearly explained or tested. To meet the checklist, add comments in the code and create tests for boundary cases to make it clear and complete.

**Review by salihakalender — CHANGES_REQUESTED** [review]
The function is well-implemented but needs clearer documentation and input validation. Unit tests are good but should include more boundary cases and validation tests for completeness.

File: solutions/tests/test_trapping_rain_water.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+import os
+import sys
+
+# Add the parent directory to the system path for importing the `trap` function.
+sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "../..")))
+
+# Import the `trap` function from the solution file
+from solutions.trapping_rain_water import trap
+
```
Comment (asagulenko) [inline solutions/tests/test_trapping_rain_water.py:20]: try 19 line like this:
from ..trapping_rain_water import trap

Comment (Mushtary) [reply solutions/tests/test_trapping_rain_water.py:20]: Tnx Anna, now it works perfectly :)

**Review by Ismatova-Rumiya — APPROVED** [review]

**Review by salihakalender — APPROVED** [review]
Your implementation demonstrates a clear understanding, good work!
